### ISSUE TITLE
[Trait] Revisit the next step of OAM catalog

### ISSUE THREAD

**Issue by resouer** [oam-dev/catalog#31]
Traits in OAM should be mostly external capabilities and higher level abstractions for built-in capabilities, not "translation" of k8s built-in api resources. I've raised similar concern in https://github.com/oam-dev/catalog/issues/26#issuecomment-645595071.

For instance, it's less valuable to create a ServiceExpose trait which simply removed label selectors - user should define k8s service as trait freely. Automatically adding a `workloadRef` is not an advantage of OAM, more like a "burden" or "compromise".

A perfect example of how a trait is defined is Ingress API v2: https://containo.us/blog/kubernetes-ingress-service-api-demystified, which should've happen in OAM community. Though we still have opportunity in other fields like `Route`, `Blue-gree/Canary`, monitoring, logging and many. Just think about it: what's still missing in Kubernetes today?

At the mean time, @zzxwill please make sure **ANY** existing k8s resource can be defined as trait/workload/scope with **zero effort**. I'd suggest modeling every application you can find in cloud native community, for example: https://github.com/istio/istio/tree/master/samples/bookinfo by OAM and put the example YAML in `catalog/applications` or somewhere else.

Also, building abstractions with zero effort is important, decomposition/composition is great and we also need to improve its UX. /cc @wonderflow

**Comment by resouer** [comment]
The similar issue also happens to https://github.com/oam-dev/catalog/tree/master/workloads.

Workloads in OAM catalog should come from wider community like Terraform modules/Crossplane, not StatefulSet, Deployment etc.

**Comment by resouer** [comment]
A related issue: whether we need to make k8s built-in capabilities work for ContainerizedWorkload? 

There are two scenarios actually:

1. The capabilities in Deployment like VolumeMounter. My suggestion is it should go with decomposition/composition.
2. The capabilities which only works for Deployment like HPA. The issue is its `targetRef`, so we could probably auto-inject childResource to it.

**Comment by resouer** [comment]
In general, there're two main features of OAM as a model:

1. Standard application definition for Kubernetes. This is very similar to Application CRD but with extra benefits:
    - A DevOps workflow with separate concerns - e.g. Operational Capabilities as Service
    - Manageability for ops capabilities (trait) - register, discover and conflict detect
        - `kubectl get traits`
        - For given component, one could directly tell out how many ops capabilities are bound to it (by yaml file, and by `kubectl tree`) 
    - More information: [kapp](https://github.com/k14s/kapp/tree/bcf08b0971684f170509d78bfbcca1bc0950e8fa#kapp), and [GC](https://github.com/kubernetes-sigs/kustomize/issues/242)
2. A framework to build abstractions (for either workload or trait).
    - composition/decomposition
    - automatic `workloadRef` (or `ownerRef`)
    - "break the abstraction" (auto-inject childResource)?
    - anything else?

**Comment by zzxwill** [comment]
As catalog is the recommended repo for OAM components/traits/applications, how about we moving proper Apps from `oam-dev/samples` to  `application` folder?

**Comment by resouer** [comment]
@zzxwill  I agree `catalog/applications` could be categories as "user stories", similar to https://github.com/tektoncd/catalog . Though we don't need to touch `oam-dev/samples` - this is where official demo apps live.

One example is knative:

```console
catagory
├── applications
│   ├── knative
│   ├── kubernetes # show off how to generate definitions and 4-5 examples of popular k8s apps
│   ├── openfaas
│   └── ...
├── workloads
│   ├── terraform # 1000+ workloads generated from Terraform modules
│   └── ...
├── traits
├── scopes

```

/cc @wonderflow
